File: ControlLight/ConfigFileCreators/ControlHardwareConfigFileCreator.cpp
#include "../ConfigCreator.h"

int main() {
	ConfigCreator builder;
	builder.RegisterSequencer(/*Id*/ 0, /*Type*/ "OpticsFoundrySequencerV1", /*IP*/ "192.168.1.112",
		/*Port*/ 57978, /*Master*/ true, /*StartDelay*/ 10, /*ClockFrequencyinMHz*/ 100,
		/*FPGAClockToBusClockRatio*/ 15, /*UseExternalClock*/ false, /*UseStrobeGenerator*/ true,
		/*UseEdgeTriggeredLatches*/ true,
		/*Connect*/ true, /*DebugOn*/ false); // 0.104 #90.108
	builder.Save();
	return 0;
	
	const int analog_out_configs[][4] = {
		{24, 1, -10, 10},
		{28, 1, -10, 10},
		{32, 1, -10, 10},
		{36, 1, -10, 10}
	};
	for (const auto& config : analog_out_configs) {
		builder.RegisterAnalogOutBoard16bit(/*Sequencer*/ 0, /*StartAddress*/ config[0],
			/*NumberChannels*/ 4, /*Signed*/ config[1] != 0, /*MinVoltage*/ config[2],
			/*MaxVoltage*/ config[3]);
	}

	for (const int address : {2, 21, 3, 4, 5, 6, 7}) {
		builder.RegisterDigitalOutBoard(/*Sequencer*/ 0, /*Address*/ address);
	}

	const int AD9854Board0ExternalClockFrequencyinMHz = 10;
	const int AD9854Board0PLLReferenceMultiplier = 20;
	builder.RegisterDDSAD9854Board(/*Version*/ 2, /*Sequencer*/ 0, /*Address*/ 128,
		/*ExternalClockFrequencyinMHz*/ AD9854Board0ExternalClockFrequencyinMHz,
		/*PLLReferenceMultiplier*/ AD9854Board0PLLReferenceMultiplier);
	builder.RegisterDDSAD9854Board(/*Version*/ 2, /*Sequencer*/ 0, /*Address*/ 132,
		/*ExternalClockFrequencyinMHz*/ AD9854Board0ExternalClockFrequencyinMHz,
		/*PLLReferenceMultiplier*/ AD9854Board0PLLReferenceMultiplier);

	for (int address = 136; address < 172; address += 4) {
		builder.RegisterDDSAD9854Board(/*Version*/ 2, /*Sequencer*/ 0, /*Address*/ address);
	}

	for (const int address : {52, 56, 60, 64, 68, 72, 76, 80, 84}) {
		builder.RegisterDDSAD9858Board(/*Sequencer*/ 0, /*Address*/ address);
	}

	for (const int address : {1, 10}) {
		builder.RegisterDDSAD9959Board(/*Sequencer*/ 0, /*Address*/ address, /*ClockFrequencyinMHz*/ 500);
	}

	builder.Save();
	return 0;
}

File: ControlLight/ConfigFileCreators/ControlHardwareConfigFileCreator.py
from ConfigCreator import ConfigBuilder


if __name__ == "__main__":
    builder = ConfigBuilder(filename="ControlHardwareConfig.json")
    builder.RegisterSequencer(IP="192.168.0.112", Port=57978, ClockFrequencyinMHz=100, StrobeDurationInFPGAClockPeriods=15, DebugOn=False)  # 0.104 #90.108
    builder.RegisterSerialPortBoard(Sequencer=0, Address=251, RackNr=0, SlotNr=1)
    analog_out_configs = [
        (24, True, -10, 10),  # each of these lines configures 4 analog outputs in consecutive order of addresses
        (28, True, -10, 10),
        (32, True, -10, 10),
        (36, True, -10, 10)
    ]
    for addr, signed, minv, maxv in analog_out_configs:
        builder.RegisterAnalogOutBoard16bit(StartAddress=addr, Signed=signed, MinVoltage=minv, MaxVoltage=maxv)

    for addr in [1, 2]:
        builder.RegisterDigitalOutBoard(Address=addr)

    AD9854Board0ExternalClockFrequencyinMHz = 10
    AD9854Board0PLLReferenceMultiplier = 20
    builder.RegisterDDSAD9854Board(Address=128, ExternalClockFrequencyinMHz=AD9854Board0ExternalClockFrequencyinMHz, PLLReferenceMultiplier=AD9854Board0PLLReferenceMultiplier)
    builder.RegisterDDSAD9854Board(Address=132, ExternalClockFrequencyinMHz=AD9854Board0ExternalClockFrequencyinMHz, PLLReferenceMultiplier=AD9854Board0PLLReferenceMultiplier)

    for addr in range(136, 172, 4):
        builder.RegisterDDSAD9854Board(Address=addr, ExternalClockFrequencyinMHz=300)

    for addr in [52, 56]:  # range(52, 84, 4):
        builder.RegisterDDSAD9858Board(Address=addr, ClockFrequencyinMHz=1200)

    for addr in [3, 4]:
        builder.RegisterDDSAD9959Board(Address=addr, ClockFrequencyinMHz=500, version = 0.09)

    builder.Save()
